waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -0,0 +1,606 @@
+            [clojure.data.json :as json]
+            [clojure.edn :as edn]
+            [clojure.java.io :as io]
+            [clojure.java.shell :as shell]
+            [clojure.string :as string]
+            [clojure.tools.logging :as log]
+            [plumbing.core :as pc]
+            [slingshot.slingshot :as ss]
+            [waiter.scheduler :as scheduler]
+            [waiter.service-description :as sd]
+            [waiter.util.date-utils :as du]
+            [waiter.util.http-utils :as http-utils]
+            [waiter.util.utils :as utils])
+  (:import (org.joda.time.format DateTimeFormat)))
+
+(defn authorization-from-environment []
+  "Sample implementation of the authentication string refresh function.
+   Returns a string to be used as the value for the Authorization HTTP header,
+   reading the string from the WAITER_K8S_AUTH_STRING environment variable."
+  (log/info "Called waiter.scheduler.kubernetes/authorization-from-environment")
+  (System/getenv "WAITER_K8S_AUTH_STRING"))
+
+(def k8s-api-auth-str
+  "Atom containing authentication string for the Kubernetes API server.
+   This value may be periodically refreshed asynchronously."
+  (atom nil))
+
+(def k8s-timestamp-format
+  "Kubernetes reports dates in ISO8061 format, sans the milliseconds component."
+  (DateTimeFormat/forPattern "yyyy-MM-dd'T'HH:mm:ss'Z'"))
+
+(defn- timestamp-str->datetime
+  "Parse a Kubernetes API timestamp string."
+  [k8s-timestamp-str]
+  (du/str-to-date k8s-timestamp-str k8s-timestamp-format))
+
+(defn- use-short-service-hash? [k8s-max-name-length]
+  ;; This is fairly arbitrary, but if we have at least 48 characters for the app name,
+  ;; then we can fit the full 32 character service-id hash, plus a hyphen as a separator,
+  ;; and still have 25 characters left for some prefix of the app name.
+  ;; If we have fewer than 48 characters, then we'll probably want to shorten the hash.
+  (< k8s-max-name-length 48))
+
+;; Kubernetes Pods have a unique 5-character alphanumeric suffix preceded by a hyphen.
+(def ^:const pod-unique-suffix-length 5)
+
+(defn- service-id->k8s-app-name [{:keys [max-name-length] :as scheduler} service-id]
+  "Shorten a full Waiter service-id to a Kubernetes-compatible application name.
+   May return the service-id unmodified if it doesn't violate the
+   configured name-length restrictions for this Kubernetes cluster."
+  (let [[_ app-prefix x y z] (re-find #"([^-]+)-(\w{8})(\w+)(\w{8})$" service-id)
+        k8s-max-name-length (- max-name-length pod-unique-suffix-length 1)
+        suffix (if (use-short-service-hash? k8s-max-name-length)
+                 (str \- x z)
+                 (str \- x y z))
+        prefix-max-length (- k8s-max-name-length (count suffix))
+        app-prefix' (cond-> app-prefix
+                      (< prefix-max-length (count app-prefix))
+                      (subs 0 prefix-max-length))]
+    (str app-prefix' suffix)))
+
+(pc/defnk replicaset->Service
```

Comment:
Interesting use of `defnk`.  I hadn't thought about using `defnk` outside of the graph-compiled stuff.  Seems nicer than using Clojure's default destructuring in this case.